---
title: "Format VCF Files"
author: "Christina Nieuwoudt, Wendy Wang, Brad McNeney, and Jinko Graham"
date: "`r Sys.Date()`"
output:
  html_document:
    number_sections: yes
    toc: yes
    toc_depth: 3
vignette: |
  %\VignetteIndexEntry{Format VCF Files} 
  %\VignetteEngine{knitr::rmarkdown} 
  %\VignetteEncoding{UTF-8}{inputenc}
---

<style type="text/css">

body{ /* Normal  */
   font-size: 12px;
}
</style>

# Introduction

ADD INTRODUCTION

NOTE: following only applicable to SNV data - no indels, cnvs, etc.




In this vigentte we demonstrate how to:

1. import a vcf file (see section XX),
2. format haplotype data (see section XX),
3. format mutation data (see section XX), 
4. create an object of class `SNVdata` (see section XX).

The data used in this vignette was obtained from 


Alternatively, we also provide pre-formatted SNV data from *... describe data (1000 Genomes, exon only, filtered for..., etc.)*.  This data can be obtained using the `import_SNVdata` function, which we demonstrate in section XX.

Describe data used in vignette

In this vignette we will illustrate how users may format exon data for chromosome 21, which is stored as a vcf file, into the appropriate format for use with the `sim_RVstudy` function.  The 

# Importing VCF files with vcfR

Importing vcf data using vcfR package [2] is accomplished with the `read.vcfR` function.  To import a vcf file the user supplies the file path to the argument `file` of the `read.vcfR` function. For example, suppose the vcf file named "exons_chr21.vcf.gz" was downloaded from https://github.com/cnieuwoudt/1000-Genomes-Exon-Data/tree/master/Exon%20Data and stored in a folder named "Data" in the "C" directory. We import "exons_chr21.vcf.gz" using the `read.vcfR` function as follows:

```{r, echo = FALSE}
load(url("https://github.com/cnieuwoudt/SimRVSequences/raw/master/data-raw/vcf_chrom21.rda"))
```
```{r}
# load the vcfR package
library(vcfR)
```
```{r, eval = FALSE}
# specify the file path
vcf_file_path <- "C:/Data/exons_chr21.vcf.gz"

#import vcf file
vcf_chrom21 <- read.vcfR(vcf_file_path)
```
```{r}
# determine the structure of vcf_chrom21
str(vcf_chrom21)
```


From the output above, we see that `vcf_chrom21` is an object of class `vcfR` with 3 slots.  The first slot `@meta` contains meta data for the vcf file; the second slot `@fix` contains the fixed data, which describes the mutations; and the third slot `@gt` contians the genotype data.  We discuss these components in detail in the following sections.  Specifically, in section section XX we discuss genotype data and in section XX we discuss the fixed data and the meta data. For additional information regarding objects of class `vcfR` execute `help("vcfR-class")` in the console.

# Format Haplotype Data{#Haps}

We now demonstrate how to convert the genotype data into haplotype data.  Recall, that the genotype data is stored in the `@gt` slot of the `vcfR` object.  

```{r}
# View the first five rows and columns of the genotype data
vcf_chrom21@gt[1:5, 1:5]
```

From the output above, we see that the first column in `vcf_chrom21@gt` does not contain genotype data, but rather a variable named `FORMAT`.  Following this variable, each column in `vcf_chrom21@gt` represents an individual identified by a unique character string, e.g. "HG00096".  Each row in `vcf_chrom21@gt` represents a SNP.

```{r}
#View first 4 mutations for the individual with ID "HG00096"
vcf_chrom21@gt[1:4, "HG00096"]
```

From the output above, we see that for the first three SNPs in the genotypes data the individual with ID "HG00096"" is homozygous for the reference allele, i.e. "0|0".  However, for the fourth SNP we see that this individual is heterozygous for the alternate allele, i.e. "0|1".  We note that since this data has been phased the genotype "0|1" indicates that the individual inherited the reference allele, "0", from his or her father and the alternate allele, "1", from his or her mother.


```{r}
# Remove the variable named "FORMAT"
# and store the resulting data as genos 
genos <- vcf_chrom21@gt[, -1]

# View the first 5 mutations (i.e. rows) for 
# the first 3 individuals (i.e. columns)
genos[1:5, 1:3]
```

To convert this genotype data to a sparseMatrix we supply the genotypes data to the `genos2sparseMatrix` function, which is included in the `SimRVSequences` package. This function may be used to convert phased genotype data for diplod organisms into a sparse matrix. We note that this conversion makes use of the methods provided with the `Matrix` package [1]. 

```{r}
# load the SimRVSequences library
library(SimRVSequences)

# Convert to sparseMatrix
Haplotypes <- genos2sparseMatrix(genotypes = genos)
```

We note that `genos2sparseMatrix` transposes the supplied genotype data.  Thus, in the returned matrix individuals are stored as rows and mutations are stored as columns. Additionally, since this is a diploid population, each individual will have two rows of data.  That is, the haplotype data for the first indivdiual in `genos`, named "HG00096", is stored in rows 1 and 2 of `Haplotypes`, the haplotype data for the second indivdiual in `genos`, named "HG00097", is stored in rows 3 and 4 of `Haplotypes`, and so on. 

```{r}
# View haplotype data for the first  
# 3 diploid individuals and first 5 SNPs
Haplotypes[1:6, 1:5]
```

From the output above we see that `Haplotypes` is a sparse matrix of class `dgCMatrix`.  Entries of `1` represent a mutated allele, while the wild type is represented as '`.`'.   For additional information regarding objects of class `dgCMatrix` execute `help("dgCMatrix-class")` in the console.

# Format Mutation Data

In addition to genotype data vcf files also contain fixed data.  Fixed data describes the SNPs in the genotype data. 

```{r}
# View the first five rows of the fixed data
vcf_chrom21@fix[1:5, ]
```

From the output above, we see that the fixed data contains several variables. According to http://www.internationalgenome.org/wiki/Analysis/vcf4.0/ the variables above may be interpreted as follows:

1. `CHROM` is the chromosome number of the SNP
2. `POS` is the position, in base pairs, of the SNP
3. `ID`, when available, a unique identifier for the SNP, e.g. rs numbers for dbSNPs.
4. `REF` is the reference allele for the SNP
5. `ALT` is the alternate allele for the SNP
6. `QUAL`, when available, is a quality score for the alternate allele
7. `FILTER` indicates whether or not the site has passed all filters.  When passing all filters `FILTER = PASS`, otherwise `FILTER` is charater string which contains all non-passing filters.
8. `INFO` contains additional information for each SNP.  Each variable in filter is separated by a semicolon.

For more information on the variables contained in `INFO` we may refer to the metadata in the vcf object.
```{r}
# View the first five rows of the fixed data
vcf_chrom21@meta
```

From the output above, we see that the first variable contained in `INFO` is named `AF` and represents the estimated allele frequency for the SNP.

To format the fixed data, including the `INFO` variables, the `vcfR` package offers the `vcfR2tidy` function.  We now demonstrate how to use the `vcfR2tidy` function to format the fixed data.  For additional information regarding the `vcfR2tidy` function please execute `help(vcfR2tidy)` in the console.

```{r}
# Extract and store the mutation data using vcfR2tidy
#
# NOTE: setting info_only = TRUE since we do not need 
# to re-process the genotype data
#
# The argument "info_fields" specify which variables to include 
# "info_types" indicates the respective variable type for each
# "info_fields" variable.
muts <- vcfR2tidy(vcf_chrom21, 
                  info_only = TRUE,
                  info_fields = c("AF", "AC", "NS", "AN",
                                  "EAS_AF", "EUR_AF", "AFR_AF", 
                                  "AMR_AF", "SAS_AF", "DP"),
                  info_types = c(AF = "n", AC = "i", NS = "i", AN = "i",
                                 EAS_AF = "n", EUR_AF = "n", AFR_AF = "n", 
                                 AMR_AF = "n", SAS_AF = "n", DP = "i"))

#determine the structure of muts
str(muts)
```

From the output above, we see that `muts` contains the fixed data as a `tbl_df` which inherits from objects of class `data.frame`.  Additionally, `muts` also contains the meta data described previously.

Our `sim_RVstudy` function expects mutation data to formatted as a data frame.  Additionally, mutation data supplied to `sim_RVstudy` must contain the following variables:

1. `colID`: a numeric variable which associates the rows in the mutations data frame to the columns in the haplotype sparse matrix. 
2. `chrom`: represents the chromosome in which the SNP resides.
3. `position`: represents the position of the SNP in base pairs.
4. `afreq`: (Optional) the derived allele frequency of the SNP.

From the previous discussion, we know that the fixed data returned by `vcf2tidy` contains the variables `chrom`, `position`, and `afreq`. To make the output from the `vcf2tidy` function compatible with the format expected for the `sim_RVstudy` function we make the following changes.

```{r}
# store the fixed item in muts data as dataframe 
Mutations <- as.data.frame(muts$fix)

# Create the variable colID to identify the column positon of the mutation.
# NOTE: Since mutations have retained their order, we can accomplish this task 
# using the seq function.  Since the mutations are stored in the columns of
# the Haplotypes matrix we determine the length of our sequence as the number of 
# columns in Haplotypes.
Mutations <- cbind(seq(1:dim(Haplotypes)[2]), Mutations)
head(Mutations)
``` 

```{r}
#rename columns for consistency with expected format
#According to muts$meta the variable "AF" is the estimated allele freq,
#i.e. the variable we expect to be named "afreq" in the Mutations dataset.
colnames(Mutations)[c(1:3, 7)] = c("colID", "chrom", "position", "afreq")
head(Mutations)
```

# Create an Object of Class `SNVdata`

Objects of class `SNVdata` inherit from a `list` and contain the following items:

1. The haplotype data; format: sparseMatrix
2. The mutation data; format: dataframe
3. sample data.

Haplotype and Mutation data are discussed 

```{r}
file_path <- 'https://raw.githubusercontent.com/cnieuwoudt/1000-Genomes-Exon-Data/master/Formatted-SNVdata/SampleData.csv'
SampleData <- read.csv(file_path)
```


```{r}
#create SNVdata object for chromosome 21 without sample or meta data
SNVdata_chrom22 <- SNVdata(Haplotypes = Haplotypes, 
                           Mutations = Mutations,
                           Samples = SampleData)

class(SNVdata_chrom22)

str(SNVdata_chrom22)
```



```{r}
#create SNVdata for chromosome 22 with sample and meta data
```



# Import 1000 Genomes Exon Data with `import_SNVdata`

The SNVdata objects for each chromosome are stored in the "Formatted-SNVdata" folder in the GitHub repository.

The `import_SNVdata` function allows users to import formatted SNV data for each chromosome.  The `import_SNVdata` function has two arguments:

  1. `chrom` A number or list of numbers that specify which non-sex chromosomes to import.
  2. `pathway_df` (Optional) A data frame that contains the positions for each exon in a pathway of interest.  This data frame must contain the variables `chrom`, `exonStart`, and `exonEnd`.  *We expect that `pathwayDF` does not contain any overlapping segments.  Users may combine overlapping exons into a single observation with our `combine_exons` function.  For additional information regarding the `combine_exons` function please  execute `help(combine_exons)` in the console.*

#### Import data for one chromosomes, chromosome 22
```{r}
#import SNV data for chromosome 21, and time
a = Sys.time()
EXdata = import_SNVdata(21)
b = Sys.time()
difftime(b, a, units = "secs")

class(EXdata)
str(EXdata)

#Note the size of "haplotypes"
dim(EXdata$Haplotypes)
```

# References
[1] Douglas Bates and Martin Maechler (2018). 
  **Matrix: Sparse and Dense Matrix Classes and Methods**. 
  *R package version 1.2-14*. https://CRAN.R-project.org/package=Matrix
  
[2] Knaus BJ, Grünwald NJ (2015). 
  **VCFR: a package to manipulate and visualize variant call format data in R.**
  Molecular Ecology Resources, 17(1): 44-53.
